src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs

```diff
@@ -569,26 +568,41 @@ private static void VerifyRefAssemblyClient(string lib_cs, string source, Action
 
         [Theory]
         [InlineData("public int M() { error(); }", true)]
-        [InlineData("public int M() { error() }", false)] // Should be true. See follow-up issue https://github.com/dotnet/roslyn/issues/17612
+        [InlineData("public int M() { error() }", false)] // This may get relaxed. See follow-up issue https://github.com/dotnet/roslyn/issues/17612
+        [InlineData("public int M();", true)]
+        [InlineData("public int M() { int Local(); }", true)]
+        [InlineData("public C();", true)]
+        [InlineData("~ C();", true)]
         [InlineData("public Error M() { return null; }", false)] // This may get relaxed. See follow-up issue https://github.com/dotnet/roslyn/issues/17612
+        [InlineData("public static explicit operator C(int i);", true)]
+        [InlineData("public async Task M();", false)]
+        [InlineData("partial void M(); partial void M();", false)] // This may get relaxed. See follow-up issue https://github.com/dotnet/roslyn/issues/17612
         public void RefAssembly_IgnoresSomeDiagnostics(string change, bool expectSuccess)
         {
             string sourceTemplate = @"
-public class C
+using System.Threading.Tasks;
+public partial class C
 {
     CHANGE
 }
 ";
-            string source = sourceTemplate.Replace("CHANGE", change);
-            string name = GetUniqueName();
-            CSharpCompilation comp1 = CreateCompilationWithMscorlib(Parse(source),
-                options: TestOptions.DebugDll.WithDeterministic(true), assemblyName: name);
+            VerifyIgnoresDiagnostics(EmitOptions.Default.WithEmitMetadataOnly(false).WithTolerateErrors(false), success: false);
+            VerifyIgnoresDiagnostics(EmitOptions.Default.WithEmitMetadataOnly(true).WithTolerateErrors(false), success: expectSuccess);
 
-            using (var output = new MemoryStream())
+            void VerifyIgnoresDiagnostics(EmitOptions emitOptions, bool success)
             {
-                var emitResult = comp1.Emit(output, options: EmitOptions.Default.WithEmitMetadataOnly(true));
-                Assert.Equal(expectSuccess, emitResult.Success);
-                Assert.Equal(!expectSuccess, emitResult.Diagnostics.Any());
+                string source = sourceTemplate.Replace("CHANGE", change);
+                string name = GetUniqueName();
+                CSharpCompilation comp = CreateCompilationWithMscorlib(Parse(source),
+                    options: TestOptions.DebugDll.WithDeterministic(true), assemblyName: name);
+
+                Assert.False(comp.DeclarationDiagnostics.HasAnyErrors());
```

Comment:
Good point.
It looks like there is an existing bug which leads to `DeclarationDiagnostics` returning an empty `DiagnosticBag` when it was not yet initialized.

```C#
  internal DiagnosticBag DeclarationDiagnostics
        {
            get
            {
                // We should only be placing diagnostics in this bag until
                // we are done gathering declaration diagnostics. Assert that is
                // the case. But since we have bugs (see https://github.com/dotnet/roslyn/issues/846)
                // we disable the assertion until they are fixed.
                Debug.Assert(!_declarationDiagnosticsFrozen || true);
                if (_lazyDeclarationDiagnostics == null)
                {
                    var diagnostics = new DiagnosticBag();
                    Interlocked.CompareExchange(ref _lazyDeclarationDiagnostics, diagnostics, null);
                }

                return _lazyDeclarationDiagnostics;
            }
        }
```